Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+
+Some of this information can be omitted or stored in more efficient form, e.g.:
+
+- The garbage collection information can be omitted for environments with conservative garbage collection, such as IL2CPP.
+- The full metadata information is not strictly required for 'private' methods or types so it is possible to strip it from the CLI image.
+- The metadata can be stored in more efficient form, such as the .NET Native metadata format.
+- The platform native executable format (ELF, Mach-O) can be used as envelope instead of PE to take advantage of platform OS loader.
+
+
+## Definition of Version Compatibility for Native Code
+
+Even for IL or unmanaged native code, there are limits to what compatible changes can be made. For example, deleting a public method is sure to be an incompatible change for any extern code using that method.
+ 
+Since CIL already has a set of [compatibility rules](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-changes.md), ideally the native format would have the same set of compatibility rules as CIL. Unfortunately, that is difficult to do efficiently in all cases. In those cases we have multiple choices
+
+1. Change the compatibility rules to disallow some changes
+2. Never generate native structures for the problematic cases (fall back to CIL techniques)
+3. Generate native structures for the problematic cases, but use them only if there was no incompatible change made
+4. Generate less efficient native code that is resilient
+ 
+Generally the hardest versioning issues revolve around
+
+- Value types (structs)
+- Generic methods over value types (structs)
+
+These are problematic because value classes are valuable precisely _because_ they have less overhead than classes. They achieve this value by being 'inlined' where they are used. This makes the code generated for value classes very fragile with respect to any changes to the value class's layout, which is bad for resilience. Generics over structs have a similar issue.
+
+Thus this proposal does _not_ suggest that we try to solve the problem of having version resilience in the presence of layout changes to value types. Instead we suggest creating a new compatibility rule
+
+- It is a breaking change to change the number or type of any (including private) fields of a public value type (struct). However if the struct is non-public (that is internal) then the restriction does not apply.
+ 
+This is a compatibility that is not present for CIL. All other changes allowed by CIL can be allowed by native code without prohibitive penalty. In particular the following changes are allowed
+
+1. Adding instance and static fields to reference classes
+2. Adding static fields to a value class. 
+3. Adding virtual, instance or static methods to a reference or value class
+4. Changing existing methods (assuming the semantics is compatible).
+5. Adding new classes.
+
+
+## Versioning Bubbles
```

Comment:
Nit: there seems to be some minor naming inconsistency here. Sometimes they're called 'versioning bubbles' and other times 'version bubbles'. We should just use one name in the documentation.